chore(login): complete login page

jniles opened the pull request:
This Pull Request addresses all the criteria in #222 and will close it when it lands in `master`.

It includes numerous bug fixes that prevent the user to bypassing the login page via url or $state manipulation.  It also prevents the user from getting back to the `/login` page by the same mechanism.  End-to-end tests prove all these cases work.  This should make our login interface bullet-proof.

Bugs have been fixed with the tree.  You can't open the tree or access the settings page when the user does not have a session.

The login page now uses the excellent `NotifyService` introduced in #370 for all server errors.  Form errors have been translated into more meaningful English, but are still displayed at the bottom of the form.  Any attempts over 3 will trigger a "fatal" error that informs the user to check with a sysadmin to reset their password.

The `SessionService` now notifies on login/logout with a welcome and goodbye message.  These can either be removed entirely if they seem tacky, or improved with interesting information like session duration, a personalized welcome citing the user's name, last login date, etc...

Finally, the user interface has been cleaned up to address issues on small screens and tested on all the chrome dev tool's responsive settings.

For a more detailed analysis, see the commit logs.

Closes #222.

---

Thank you for contributing!

Before submitting this pull request, please verify that you have:
- [x] Run your code through JSHint.  [Check out our styleguide](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/docs/STYLEGUIDE.md).
- [x] Run integration tests.
- [x] Run end-to-end tests.
- [x] Accurately described the changes your are making in this pull request.

For a more detailed checklist, [see the official review checklist](https://docs.google.com/document/d/1nupLVLRXgSZJQo_acLgrwvPnN8RukfSiwRhSToj81uU/pub) that this PR will be evaluated against.

Ensuring that the above checkboxes are completed will help speed the review process and help build a stronger application.  Thanks!

jniles committed:
fix(login): disable buttons w/o session

This commit addresses two issues reported in #222:
 1. The duplicate text on the login panel has been removed.
 2. The tree cannot be expanded when the user does not have a session.

jniles committed:
feat(login): NotifyService for login messages

This commit cleans up the auth/login messages by moving server-sent
error messages into `growl-notification` service.  It also exposes a new
method on the `NotifyService` called `fatal()` to signal fatal errors
using the same colorscheme as `NotifyService.handleError()`.  The name
fatal was chosen to encourage developers NOT to use this, except in
very rare cases.  The `danger()` method should be preferred in almost
all scenarios.

This commit also adds the `NotifyService` into the `SessionService` to
display welcome and goodbye messages on login/logout events.

This brings us closer to finishing #222, closing all error handling and
text-related issues.

jniles committed:
fix(login): ensure $locationChange respects login

This commit fixes a critical bug introduced throughout the `ui-router`
upgrade that allows users to access any page without logging in to the
application.  This has been patched so that $locationChange does not
even begin if the user session does not exist.

Finally, informative messages have been added to tell the user exactly
what is going on and what to do next when the session exists, doesn't
exist, or otherwise.

jniles committed:
fix(login): prevent $state.go() getting to login

This commit ensures that $state.go() doesn't get to `/login` when the
user already has a session logged in.  Since $state-based and URL-based
routing are decoupled in bhima, we need to two checks to make sure the
URL is correctly displayed (and doesn't transition) and the $state is
correctly blocked and doesn't transition.

Fixes bugs reported in #222.

jniles committed:
feature(login): polish login to 2.x standards

The login procedure has been completely polished with notifications,
better error handling, translated text, and a centered panel.  All bugs
related to the tree and/or navigation have been squashed.  This is ready
for a 2.x review and approval by the team.

End-to-end tests have been written to future-proof the behavior of the
module.  Tests have been written for every case discussed in #222 and
bugs found along the way.

The breakpoints have been adjusted so that the login form nicely renders
on all default responsive devices included in the chrome dev tools.

Decisions, such as welcome messages, should be reviewed by the team.

Closes #222.

jniles committed:
fix(test): fix unit tests for SessionService

This commit fixes the unit tests for SessionService by injecting
`$translate`'s module into Karma's namespace.

sfount commented on client/src/index.html:
```diff
@@ -92,10 +92,9 @@
 
     <!-- System wide notifications -->
     <div class="ng-cloak">
-      <growl-notification data-success-notification ng-repeat="notification in AppCtrl.notifications" ng-class="notification.format" class="fading ng-animate-enabled" ttl="notification.ttl">
```
Good catch! Certainly not always true.

sfount commented on client/src/partials/login/login.js:
```diff
@@ -77,12 +74,14 @@ function LoginController(AppCache, Session, Languages, Projects) {
+      // if the user has tried too many times, display a fatal error working for
+      // ten seconds.
+      if (maxAttempts <= attempts++) {
+        return Notify.fatal('AUTH.TOO_MANY_TRYS', 10000);
```
I would reserve the use of the `fatal` style for errors that are beyond the users control, this would mean that whenever you see dark red it probably means a system administrator has to be involved. Whilst this use case is could be argued to be that way the user can still just put in their correct password. 

What do you think?

Too many attempts could potentially ramp up from `warn` to `danger` although I wouldn't have a problem with it being `danger` uniformly.

jniles replied:
No issue with that logic.  I'll make the change to use `danger` uniformly
with both types or errors.

jniles committed:
chore: remove fatal method from Notify

This commit removes the `fatal()` method from the Notify service - the
`fatal` messages should be reserved for errors that are out of reach of
the user and cannot be fixed without a sysadmin.  These messages now use
`danger` to render uniform errors.

It also removes welcome/goodbye messages from the SessionService.
